### PR TITLE
Use serial queue to stop AudioUnits when handling errors in a render callback.

### PR DESCRIPTION
**Pull request by kinetiknz**
Avoids a possible deadlock inside CoreAudio if another thread calls AudioOutputUnitStop concurrently.  The thread running the render callback holds the HALB_Mutex and blocks acquiring the CAMutex inside AudioOutputUnitStop. The racing thread calling AudioOutputUnitStop holds the CAMutex and blocks acquiring the HALB_Mutex.

### CODE REVIEW

File: src/backend/mod.rs [hunk 1]
```diff
@@ -669,9 +677,17 @@ extern "C" fn audiounit_output_callback(
 
     if outframes < 0 || outframes > i64::from(output_frames) {
         stm.stopped.store(true, Ordering::SeqCst);
-        stm.core_stream_data.stop_audiounits();
-        audiounit_make_silent(&mut buffers[0]);
         stm.notify_state_changed(State::Error);
```
Comment (ChunMinChang) [inline src/backend/mod.rs:680]: Should the state notification be fired after silencing the output buffer?

**Review by ChunMinChang — COMMENTED** [review]
I'd like to figure out what happen exactly first. So there are two threads. One holds *HALB_Mutex* and the other holds *CAMutex*. I'd like to know what thread, or which call, holds or tries to acquire *HALB_Mutex* and  *CAMutex*. Does the thread running input or output callback hold the *HALB_Mutex*? Does calling `stop_audiounits` lead to *CAMutex* acquisition? Is it possible to write a test to simulate it?

**Comment by kinetiknz** [comment]
Gecko thread stack:
```
 0  libsystem_kernel.dylib!__psynch_mutexwait
 1  libsystem_pthread.dylib!_pthread_mutex_firstfit_lock_slow
 2  CoreAudio!HALB_Mutex::Lock
 3  CoreAudio!HALC_ProxyIOContext::StopIOProc
 4  CoreAudio!HAL_HardwarePlugIn_DeviceStop
 5  CoreAudio!HALDevice::StopIOProc
 6  CoreAudio!AudioDeviceStop
 7  CoreAudio
 8  CoreAudio
 9  CoreAudio
10  XUL!coreaudio_sys_utils::audio_unit::audio_output_unit_stop
11  XUL!cubeb_coreaudio::backend::stop_audiounit
12  XUL!cubeb_coreaudio::backend::CoreStreamData::stop_audiounits
13  XUL!<cubeb_coreaudio::backend::AudioUnitStream as cubeb_backend::traits::StreamOps>::stop::{{closure}}
```

CoreAudio render thread (com.apple.audio.IOThread.client):
```
 0  libsystem_kernel.dylib!__psynch_mutexwait
 1  libsystem_pthread.dylib!_pthread_mutex_firstfit_lock_slow
 2  libAudioToolboxUtility.dylib!CAMutex::Lock
 3  CoreAudio
 4  XUL!coreaudio_sys_utils::audio_unit::audio_output_unit_stop
 5  XUL!cubeb_coreaudio::backend::stop_audiounit
 6  XUL!cubeb_coreaudio::backend::CoreStreamData::stop_audiounits
 7  XUL!cubeb_coreaudio::backend::audiounit_output_callback
 8  CoreAudio
 9  CoreAudio
10  CoreAudio
11  CoreAudio!invocation function for block in HALC_ProxyIOContext::HALC_ProxyIOContext
12  CoreAudio!HALB_IOThread::Entry
```

CoreAudio's render thread holds HALB_Mutex during our audiounit_output_callback, then blocks on CAMutex.  The Gecko thread holds CAMutex (inside AudioOutputUnitStop), then blocks on HALB_Mutex.

It'd be quite difficult to write a useful test for this as the the incorrect code would still pass whenever the stop from inside the render callback completed the CAMutex section of its work before the stop from another thread acquires CAMutex.

File: src/backend/mod.rs (hunk 1 above)
Comment (kinetiknz) [reply src/backend/mod.rs:680]: The buffer wouldn't be used until we return from the render callback, so I can't see a reason for the ordering to matter.  Can you expand on your concern here?

Comment (ChunMinChang) [reply src/backend/mod.rs:680]: It seems the order shouldn't make a difference

**Review by ChunMinChang — APPROVED** [review]

**Pull request merged**